{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3b1b8be4-9f67-4145-9a1c-f0625db374d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTRL_1\n",
      "CTRL_2\n",
      "CTRL_2D\n",
      "GEM_2\n",
      "TGFb1_1\n",
      "TGFb1_2\n",
      "TGFb1_GEM_1\n",
      "TGFb1_GEM_2\n"
     ]
    }
   ],
   "source": [
    "## file folder\n",
    "import os\n",
    "\n",
    "folder_path = '/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs'\n",
    "\n",
    "# Get the list of files and directories in the folder\n",
    "folder_content = os.listdir(folder_path)\n",
    "\n",
    "# Print the content of the folder\n",
    "for item in folder_content:\n",
    "    print(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8c9ada8-fce8-4c69-a240-5a337aaab8d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bbda7771-0358-48b1-82ab-ff899531408f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/CTRL_1/count/sample_filtered_feature_bc_matrix\n",
      "/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/CTRL_2/count/sample_filtered_feature_bc_matrix\n",
      "/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/CTRL_2D/count/sample_filtered_feature_bc_matrix\n",
      "/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/GEM_2/count/sample_filtered_feature_bc_matrix\n",
      "/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/TGFb1_1/count/sample_filtered_feature_bc_matrix\n",
      "/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/TGFb1_2/count/sample_filtered_feature_bc_matrix\n",
      "/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/TGFb1_GEM_1/count/sample_filtered_feature_bc_matrix\n",
      "/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/TGFb1_GEM_2/count/sample_filtered_feature_bc_matrix\n"
     ]
    }
   ],
   "source": [
    "def unzip_all_gz_files(folder_path):\n",
    "    for file_name in os.listdir(folder_path):\n",
    "        file_path = os.path.join(folder_path, file_name)\n",
    "        \n",
    "\n",
    "        if file_name.endswith('.gz'):\n",
    "            unzipped_file_path = os.path.splitext(file_path)[0]  # Remove the .gz extension\n",
    "            #print(file_path)\n",
    "            with gzip.open(file_path, 'rb') as gz_file:\n",
    "                with open(unzipped_file_path, 'wb') as unzipped_file:\n",
    "                    unzipped_file.write(gz_file.read())\n",
    "            \n",
    "            # Optionally, you can delete the .gz file after unzipping\n",
    "            # os.remove(file_path)\n",
    "\n",
    "\n",
    "samples = folder_content\n",
    "for sample in samples:\n",
    "\n",
    "    folder_path = f'/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/{sample}/count/sample_filtered_feature_bc_matrix'\n",
    "    df = unzip_all_gz_files(folder_path)\n",
    "\n",
    "    #print(df.head())\n",
    "    print(folder_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "55ab1f2a-c1a7-425c-a55e-ee7cdfead8fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices\n",
    "import scipy.io as sio\n",
    "import os\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8bf9e49f-0a7b-4e04-8623-ed6475f6f95a",
   "metadata": {},
   "outputs": [],
   "source": [
    "##samples = ['sample1', 'sample2']  # Add more sample names if needed\n",
    "conditions = folder_content\n",
    "#pwd /storage/users/sac43cg/res_Samantha/outs/per_sample_outs/CTRL_1/count/sample_filtered_feature_bc_matrix\n",
    "\n",
    "# Create an empty list to store the AnnData objects\n",
    "adata_list = []\n",
    "barcodes_list = []\n",
    "features_list = []\n",
    "matrix_list = []\n",
    "#combined_adata = []\n",
    "\n",
    "for condition in conditions:\n",
    "           \n",
    "    # Define the file paths\n",
    "    base_path = f'/storage/users/sac43cg/res_Samantha_1/outs/per_sample_outs/{condition}/count/sample_filtered_feature_bc_matrix/'  # Replace with the actual path to your data\n",
    "\n",
    "    barcodes_file = os.path.join(base_path, 'barcodes.tsv')\n",
    "    features_file = os.path.join(base_path, 'features.tsv')\n",
    "    matrix_file = os.path.join(base_path, 'matrix.mtx')\n",
    "\n",
    "    # Read the barcodes.tsv file\n",
    "    barcodes = pd.read_csv(barcodes_file, header=None, index_col=0, names=['barcode'])\n",
    "\n",
    "    # Read the features.tsv file\n",
    "    features = pd.read_csv(features_file, sep='\\t', header=None, index_col=0, names=['feature_id', 'symbol', 'type'])\n",
    "    # get only gene expression features\n",
    "    #features = features[features['type'] == 'gene expression']\n",
    "\n",
    "    # Read the matrix.mtx file\n",
    "    matrix = sio.mmread(matrix_file)\n",
    "\n",
    "    # Create the AnnData object\n",
    "    adata = ad.AnnData(X=matrix.T, obs=barcodes, var=features)\n",
    "    \n",
    "    # Add a column to the observation metadata to indicate the condition\n",
    "    adata.obs['condition'] = condition\n",
    "\n",
    "    # Append the AnnData object to the list\n",
    "    adata_list.append(adata)\n",
    "    features['condition'] = condition\n",
    "    features_list.append(features)\n",
    "        \n",
    "    \n",
    "# Combine all the AnnData objects into a single AnnData object\n",
    "combined_adata = ad.concat(adata_list, join='outer')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c3035a68-2e30-4133-b2c8-6cb1bb9185cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_adata.write('/storage/users/sac43cg/res_Samantha_1/outs/h5as_files/combined_adata.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "279ddc27-333d-49d0-a392-1e1ae746be36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 11919 × 36613\n",
       "    obs: 'condition'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "703915f7-d27d-4a3e-b3ef-5b1f3255f6ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count matrix:\n",
      "  (0, 29)\t3\n",
      "  (0, 58)\t1\n",
      "  (0, 59)\t2\n",
      "  (0, 62)\t4\n",
      "  (0, 73)\t1\n",
      "  (0, 86)\t2\n",
      "  (0, 94)\t1\n",
      "  (0, 97)\t6\n",
      "  (0, 99)\t1\n",
      "  (0, 104)\t4\n",
      "  (0, 105)\t1\n",
      "  (0, 132)\t1\n",
      "  (0, 142)\t1\n",
      "  (0, 169)\t2\n",
      "  (0, 170)\t46\n",
      "  (0, 190)\t1\n",
      "  (0, 194)\t1\n",
      "  (0, 202)\t2\n",
      "  (0, 208)\t2\n",
      "  (0, 211)\t2\n",
      "  (0, 216)\t1\n",
      "  (0, 219)\t11\n",
      "  (0, 227)\t1\n",
      "  (0, 238)\t2\n",
      "  (0, 241)\t1\n",
      "  :\t:\n",
      "  (11918, 36432)\t1\n",
      "  (11918, 36433)\t3\n",
      "  (11918, 36443)\t1\n",
      "  (11918, 36513)\t1\n",
      "  (11918, 36559)\t22\n",
      "  (11918, 36560)\t21\n",
      "  (11918, 36561)\t27\n",
      "  (11918, 36562)\t50\n",
      "  (11918, 36563)\t11\n",
      "  (11918, 36564)\t21\n",
      "  (11918, 36565)\t27\n",
      "  (11918, 36566)\t64\n",
      "  (11918, 36567)\t2\n",
      "  (11918, 36568)\t43\n",
      "  (11918, 36569)\t9\n",
      "  (11918, 36570)\t2\n",
      "  (11918, 36571)\t14\n",
      "  (11918, 36601)\t76\n",
      "  (11918, 36602)\t103\n",
      "  (11918, 36604)\t46\n",
      "  (11918, 36605)\t509\n",
      "  (11918, 36606)\t41\n",
      "  (11918, 36607)\t22\n",
      "  (11918, 36608)\t6096\n",
      "  (11918, 36609)\t801\n",
      "\n",
      "Observations (Cells):\n",
      "                      condition\n",
      "barcode                        \n",
      "AAACCCAGTAGGCTCC-1       CTRL_1\n",
      "AAACCCAGTCGCAACC-1       CTRL_1\n",
      "AAACCCATCCACAAGT-1       CTRL_1\n",
      "AAACGCTTCACTACGA-1       CTRL_1\n",
      "AAACGCTTCAGCCTCT-1       CTRL_1\n",
      "...                         ...\n",
      "TTTACTGTCATTACTC-1  TGFb1_GEM_2\n",
      "TTTCACAGTGATATAG-1  TGFb1_GEM_2\n",
      "TTTCACATCGATGCAT-1  TGFb1_GEM_2\n",
      "TTTCATGTCCCTCTAG-1  TGFb1_GEM_2\n",
      "TTTGACTGTTCCGGTG-1  TGFb1_GEM_2\n",
      "\n",
      "[11919 rows x 1 columns]\n",
      "\n",
      "Variables (Genes):\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945, ENSG00000239906, ENSG00000241860, ENSG00000241599, ENSG00000286448, ENSG00000236601, ENSG00000284733, ENSG00000235146, ENSG00000284662, ENSG00000229905, ENSG00000237491, ENSG00000177757, ENSG00000228794, ENSG00000225880, ENSG00000230368, ENSG00000272438, ENSG00000230699, ENSG00000241180, ENSG00000223764, ENSG00000187634, ENSG00000188976, ENSG00000187961, ENSG00000187583, ENSG00000187642, ENSG00000272512, ENSG00000188290, ENSG00000187608, ENSG00000224969, ENSG00000188157, ENSG00000242590, ENSG00000273443, ENSG00000237330, ENSG00000131591, ENSG00000285812, ENSG00000223823, ENSG00000272141, ENSG00000205231, ENSG00000162571, ENSG00000186891, ENSG00000186827, ENSG00000078808, ENSG00000176022, ENSG00000184163, ENSG00000260179, ENSG00000160087, ENSG00000230415, ENSG00000162572, ENSG00000131584, ENSG00000169972, ENSG00000127054, ENSG00000240731, ENSG00000224051, ENSG00000169962, ENSG00000107404, ENSG00000162576, ENSG00000175756, ENSG00000221978, ENSG00000224870, ENSG00000242485, ENSG00000272455, ENSG00000235098, ENSG00000225905, ENSG00000205116, ENSG00000225285, ENSG00000179403, ENSG00000215915, ENSG00000160072, ENSG00000197785, ENSG00000205090, ENSG00000160075, ENSG00000215014, ENSG00000228594, ENSG00000286989, ENSG00000272106, ENSG00000197530, ENSG00000189409, ENSG00000248333, ENSG00000272004, ENSG00000189339, ENSG00000008128, ENSG00000215790, ENSG00000008130, ENSG00000078369, ENSG00000231050, ENSG00000169885, ENSG00000178821, ENSG00000142609, ENSG00000233542, ENSG00000187730, ENSG00000226969, ENSG00000067606, ENSG00000271806, ENSG00000182873, ENSG00000162585, ENSG00000234396, ENSG00000157933, ...]\n",
      "\n",
      "[36613 rows x 0 columns]\n"
     ]
    }
   ],
   "source": [
    "# View the count matrix\n",
    "print(\"Count matrix:\")\n",
    "print(combined_adata.X)\n",
    "\n",
    "# View the observations (cells) information\n",
    "print(\"\\nObservations (Cells):\")\n",
    "print(combined_adata.obs)\n",
    "\n",
    "# View the variables (genes) information\n",
    "print(\"\\nVariables (Genes):\")\n",
    "print(combined_adata.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1d869f9e-936c-4200-8d99-726b6dd20213",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Observation annotations:\n",
      "                   condition\n",
      "barcode                     \n",
      "AAACCCAGTAGGCTCC-1    CTRL_1\n",
      "AAACCCAGTCGCAACC-1    CTRL_1\n",
      "AAACCCATCCACAAGT-1    CTRL_1\n",
      "AAACGCTTCACTACGA-1    CTRL_1\n",
      "AAACGCTTCAGCCTCT-1    CTRL_1\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "\n",
      "Variable annotations:\n",
      "AxisArrays with keys: \n",
      "\n",
      "Variable features:\n",
      "                      symbol             type    condition\n",
      "feature_id                                                \n",
      "ENSG00000243485  MIR1302-2HG  Gene Expression  TGFb1_GEM_2\n",
      "ENSG00000237613      FAM138A  Gene Expression  TGFb1_GEM_2\n",
      "ENSG00000186092        OR4F5  Gene Expression  TGFb1_GEM_2\n",
      "ENSG00000238009   AL627309.1  Gene Expression  TGFb1_GEM_2\n",
      "ENSG00000239945   AL627309.3  Gene Expression  TGFb1_GEM_2\n",
      "['Gene Expression' 'Multiplexing Capture']\n",
      "\n",
      "Number of observations (cells): 11919\n",
      "Number of variables (features): 36613\n"
     ]
    }
   ],
   "source": [
    "##### OPTIONAL ######\n",
    "# here you can check your generated h5ad\n",
    "# Print the annotations\n",
    "print(\"Observation annotations:\")\n",
    "print(combined_adata.obs.head())  # Print the first few rows of observation annotations\n",
    "\n",
    "print(\"\\nVariable annotations:\")\n",
    "print(combined_adata.var.head())  # Print the first few rows of variable annotations\n",
    "\n",
    "print(\"\\nVariable annotations:\")\n",
    "print(combined_adata.varm)  # Print the first few rows of variable annotations\n",
    "\n",
    "print(\"\\nVariable features:\")\n",
    "print(features.head())\n",
    "print(features['type'].unique())\n",
    "\n",
    "# Print the dimensions\n",
    "print(\"\\nNumber of observations (cells):\", combined_adata.n_obs)\n",
    "print(\"Number of variables (features):\", combined_adata.n_vars)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d9c85546-7976-410b-8af8-1647bc45d424",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Observation annotations:\n",
      "                   condition\n",
      "barcode                     \n",
      "AAACCCAGTAGGCTCC-1    CTRL_1\n",
      "AAACCCAGTCGCAACC-1    CTRL_1\n",
      "AAACCCATCCACAAGT-1    CTRL_1\n",
      "AAACGCTTCACTACGA-1    CTRL_1\n",
      "AAACGCTTCAGCCTCT-1    CTRL_1\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "Condition: CTRL_1\n",
      "\n",
      "Number of features: 36613\n",
      "\n",
      "Number of barcodes: 2973\n",
      "\n",
      "\n",
      "Observation annotations:\n",
      "                   condition\n",
      "barcode                     \n",
      "AAAGGATGTAGGAGTC-1    CTRL_2\n",
      "AAAGGGCCACATACGT-1    CTRL_2\n",
      "AAAGGGCGTTGAGAGC-1    CTRL_2\n",
      "AAAGGTACAGAGCTAG-1    CTRL_2\n",
      "AAAGGTAGTAGAGACC-1    CTRL_2\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "Condition: CTRL_2\n",
      "\n",
      "Number of features: 36613\n",
      "\n",
      "Number of barcodes: 450\n",
      "\n",
      "\n",
      "Observation annotations:\n",
      "                   condition\n",
      "barcode                     \n",
      "AAACCCAAGAGGTATT-1   CTRL_2D\n",
      "AAACCCAAGGATCACG-1   CTRL_2D\n",
      "AAACCCAAGGGCCCTT-1   CTRL_2D\n",
      "AAACCCAGTGCCCAGT-1   CTRL_2D\n",
      "AAACGAAAGATTCGCT-1   CTRL_2D\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "Condition: CTRL_2D\n",
      "\n",
      "Number of features: 36613\n",
      "\n",
      "Number of barcodes: 4819\n",
      "\n",
      "\n",
      "Observation annotations:\n",
      "                   condition\n",
      "barcode                     \n",
      "AAAGTCCAGATTACCC-1     GEM_2\n",
      "AAAGTCCAGCTTCATG-1     GEM_2\n",
      "AAAGTGACAGACACAG-1     GEM_2\n",
      "AACAAAGAGGTATCTC-1     GEM_2\n",
      "AACAACCTCACACGAT-1     GEM_2\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "Condition: GEM_2\n",
      "\n",
      "Number of features: 36613\n",
      "\n",
      "Number of barcodes: 422\n",
      "\n",
      "\n",
      "Observation annotations:\n",
      "                   condition\n",
      "barcode                     \n",
      "AAACCCACACTTTATC-1   TGFb1_1\n",
      "AAACCCACAGCTGGTC-1   TGFb1_1\n",
      "AAACCCATCTCGTTTA-1   TGFb1_1\n",
      "AAACGAAAGAGTTCGG-1   TGFb1_1\n",
      "AAACGAAAGTAATTGG-1   TGFb1_1\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "Condition: TGFb1_1\n",
      "\n",
      "Number of features: 36613\n",
      "\n",
      "Number of barcodes: 2025\n",
      "\n",
      "\n",
      "Observation annotations:\n",
      "                   condition\n",
      "barcode                     \n",
      "AAACGCTCAGTTTCAG-1   TGFb1_2\n",
      "AAAGAACGTCAAGTTC-1   TGFb1_2\n",
      "AAAGGGCTCGCCAGTG-1   TGFb1_2\n",
      "AAAGGTATCAAGGACG-1   TGFb1_2\n",
      "AAAGTCCCAACCTATG-1   TGFb1_2\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "Condition: TGFb1_2\n",
      "\n",
      "Number of features: 36613\n",
      "\n",
      "Number of barcodes: 359\n",
      "\n",
      "\n",
      "Observation annotations:\n",
      "                      condition\n",
      "barcode                        \n",
      "AAACGAATCGGCTATA-1  TGFb1_GEM_1\n",
      "AACAAGACACGTCGGT-1  TGFb1_GEM_1\n",
      "AACAAGATCTCAGTCC-1  TGFb1_GEM_1\n",
      "AACCAACAGCGGATCA-1  TGFb1_GEM_1\n",
      "AACGTCAGTGGTTCTA-1  TGFb1_GEM_1\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "Condition: TGFb1_GEM_1\n",
      "\n",
      "Number of features: 36613\n",
      "\n",
      "Number of barcodes: 386\n",
      "\n",
      "\n",
      "Observation annotations:\n",
      "                      condition\n",
      "barcode                        \n",
      "AAACCCATCAACGTGT-1  TGFb1_GEM_2\n",
      "AAACGCTAGGAGTACC-1  TGFb1_GEM_2\n",
      "AAAGGTAAGGTATAGT-1  TGFb1_GEM_2\n",
      "AAAGTCCCATGTGCCG-1  TGFb1_GEM_2\n",
      "AACCAACAGCCATCCG-1  TGFb1_GEM_2\n",
      "\n",
      "Variable annotations:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945]\n",
      "Condition: TGFb1_GEM_2\n",
      "\n",
      "Number of features: 36613\n",
      "\n",
      "Number of barcodes: 485\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Iterate over each condition\n",
    "for condition in combined_adata.obs['condition'].unique():\n",
    "    condition_adata = combined_adata[combined_adata.obs['condition'] == condition]\n",
    "    \n",
    "    # Get the number of features and barcodes for the current condition\n",
    "    num_features = condition_adata.n_vars\n",
    "    num_barcodes = condition_adata.n_obs\n",
    "    \n",
    "    print(\"Observation annotations:\")\n",
    "    print(condition_adata.obs.head())  # Print the first few rows of observation annotations\n",
    "    \n",
    "    print(\"\\nVariable annotations:\")\n",
    "    print(condition_adata.var.head())  # Print the first few rows of variable annotations\n",
    "    \n",
    "    # Write the condition name, number of features, and number of barcodes to the file\n",
    "    print(f\"Condition: {condition}\\n\")\n",
    "    print(f\"Number of features: {num_features}\\n\")\n",
    "    print(f\"Number of barcodes: {num_barcodes}\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2b82c43f-599c-4920-aa9e-18b0c69ebb20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " (485, 0)\n",
      "\n",
      " (36613, 3)\n",
      "\n",
      " (11919, 36613)\n",
      "\n",
      " (36613, 485)\n"
     ]
    }
   ],
   "source": [
    "dimensions = barcodes.shape\n",
    "print('\\n' , dimensions)\n",
    "dimensions = features.shape\n",
    "print('\\n' , dimensions)\n",
    "dimensions = combined_adata.shape\n",
    "print('\\n' , dimensions)\n",
    "dimensions = matrix.shape\n",
    "print('\\n' , dimensions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dea9696-54b2-4c42-8ad3-a2718af0a329",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5122ca97-fc8f-41e1-a14a-fd19f7a3fef3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 11919 × 36613\n",
      "    obs: 'condition'\n"
     ]
    }
   ],
   "source": [
    "print(combined_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d217927-8f95-479e-8d10-51a84ecf7482",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Gudruns_PDAC",
   "language": "python",
   "name": "gudruns_pdac"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
